File: app/src/main/java/com/stephenmorgandevelopment/thelinuxmanual/utils/Preferences.java
package com.stephenmorgandevelopment.thelinuxmanual.utils;

import android.annotation.SuppressLint;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.preference.PreferenceManager;

import com.stephenmorgandevelopment.thelinuxmanual.R;
import com.stephenmorgandevelopment.thelinuxmanual.distros.UbuntuHtmlAdapter;

import java.util.ResourceBundle;

public class Preferences {
    private static final String RELEASE = "RELEASE";
//    private static final String COLOR_PRIMARY = "COLOR_PRIMARY";
//    private static final String COLOR_SECONDARY = "COLOR_SECONDARY";
//    private static final String COLOR_THIRD = "COLOR_THIRD";
//    private static final String COLOR_ACCENT = "COLOR_ACCENT";

    private static final String LIST_FONT_SIZE = "LIST_FONT_SIZE";
    private static final String DETAILS_FONT_SIZE = "DETAILS_FONT_SIZE";

    @SuppressLint("ApplySharedPref")
    public static void setRelease(String release) {
        SharedPreferences.Editor editor = PreferenceManager.getDefaultSharedPreferences(Helpers.getApplicationContext()).edit();
        editor.putString(RELEASE, release).commit();
    }

    public static String getRelease() {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(Helpers.getApplicationContext());
        return prefs.getString(RELEASE, UbuntuHtmlAdapter.Release.FOCAL.getName());
    }

    public static int getListFontSize() {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(Helpers.getApplicationContext());
        return prefs.getInt(LIST_FONT_SIZE, 16);
    }

    public static int getDetailsFontSize() {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(Helpers.getApplicationContext());
        return prefs.getInt(DETAILS_FONT_SIZE, 16);
    }

    @SuppressLint("ApplySharedPref")
    public static void setListFontSize(int size) {
        SharedPreferences.Editor editor = PreferenceManager.getDefaultSharedPreferences(Helpers.getApplicationContext()).edit();
        editor.putInt(LIST_FONT_SIZE, size).commit();
    }

    @SuppressLint("ApplySharedPref")
    public static void setDetailsFontSize(int size) {
        SharedPreferences.Editor editor = PreferenceManager.getDefaultSharedPreferences(Helpers.getApplicationContext()).edit();
        editor.putInt(LIST_FONT_SIZE, size).commit();
    }
}

File: app/src/main/java/com/stephenmorgandevelopment/thelinuxmanual/utils/Preferences.kt
@file:Suppress("DEPRECATION")
@file:SuppressLint("UseKtx")
@file:OptIn(DelicateCoroutinesApi::class)

package com.stephenmorgandevelopment.thelinuxmanual.utils

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.stephenmorgandevelopment.thelinuxmanual.distros.ubuntu.AvailableReleases.NOBLE
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

data class CurrentPreferences(
    val release: String,
    val tabsOnBottom: Boolean,
    val searchOnBottm: Boolean,
)

@Singleton
class Preferences @Inject constructor(
    @ApplicationContext private val applicationContext: Context,
) {
    private val currentPrefs get() = CurrentPreferences(release, tabsOnBottom, searchOnBottom)
    private val _preferenceListener = MutableStateFlow(currentPrefs)
    val preferenceListener = _preferenceListener.asStateFlow()

    /**
     * Added due to issues that arise in the http/database creation and usage,
     * when having to wait for the shared prefs to come from the disk.
     */
    val currentRelease
        get() = preferenceListener.value.release

    private val preferences: SharedPreferences?
        get() = try {
            PreferenceManager.getDefaultSharedPreferences(applicationContext)
        } catch (e: Throwable) {
            null
        }

    var release: String
        private set(value) {
            preferences?.edit()?.putString(RELEASE, value)?.commit().also {
                if (it == true) _preferenceListener.tryEmit(currentPrefs)
            } ?: javaClass.wlog("Failed storing release value in shared prefs.")
        }
        get() {
            return (preferences?.getString(RELEASE, NOBLE.pathString)
                ?: NOBLE.pathString)
        }

    var tabsOnBottom: Boolean
        private set(value) {
            preferences?.edit()?.putBoolean(TABS_ON_BOTTOM, value)?.commit().also {
                if (it == true) _preferenceListener.tryEmit(currentPrefs)
            } ?: javaClass.wlog("Failed to store release value in shared prefs.")
        }
        get() {
            return preferences?.getBoolean(TABS_ON_BOTTOM, false) == true
        }

    var searchOnBottom: Boolean
        private set(value) {
            preferences?.edit()?.putBoolean(SEARCH_ON_BOTTOM, value)?.commit().also {
                if (it == true) _preferenceListener.tryEmit(currentPrefs)
            } ?: javaClass.wlog("Failed to store release value in shared prefs.")
        }
        get() {
            return preferences?.getBoolean(SEARCH_ON_BOTTOM, false) == true
        }

    inner class PreferencesWriteAccess() {
        fun setRelease(versionRelease: String) = GlobalScope.async {
            withContext(Dispatchers.IO) {
                release = versionRelease
            }
        }

        fun setTabsOnBottom(enabled: Boolean) = GlobalScope.async {
            withContext(Dispatchers.IO) {
                tabsOnBottom = enabled
            }
        }

        fun setSearchOnBottom(enabled: Boolean) = GlobalScope.async {
            withContext(Dispatchers.IO) {
                searchOnBottom = enabled
            }
        }
    }

    companion object {
        private const val LOG_TAG = "UbuntuManPages"
        private const val RELEASE: String = "RELEASE"
        private const val TABS_ON_BOTTOM: String = "TABS_ON_BOTTOM"
        private const val SEARCH_ON_BOTTOM: String = "SEARCH_ON_BOTTOM"
    }
}